File: Medium/2025-12-10-3577-Count-the-Number-of-Computer-Unlocking-Permutations/solution.py
class Solution:
    def countOrders(self, complexity: List[int]) -> int:
        MOD = 10**9 + 7
        n = len(complexity)
        
        # Sort the complexity with their original indices
        indexed_complexity = sorted((c, i) for i, c in enumerate(complexity))
        
        # dp[i] will store the number of valid permutations for the first i computers
        dp = [0] * (n + 1)
        dp[0] = 1  # Base case: 1 way to unlock 0 computers
        
        # Count how many computers have the same complexity
        count = [0] * (n + 1)
        
        for i in range(1, n + 1):
            c, idx = indexed_complexity[i - 1]
            count[i] = count[i - 1]
            # Count how many computers have complexity less than the current
            for j in range(i):
                if indexed_complexity[j][0] < c:
                    count[i] += 1
            
            # Calculate the number of valid permutations
            dp[i] = dp[i - 1] * count[i] % MOD
        
        return dp[n]

File: Medium/2025-12-10-3577-Count-the-Number-of-Computer-Unlocking-Permutations/solution.java
import java.util.Arrays;

public class Solution {
    private static final int MOD = 1000000007;

    public int countOrders(int[] complexity) {
        int n = complexity.length;
        Arrays.sort(complexity);
        
        long[] dp = new long[n + 1];
        dp[0] = 1; // Base case: 1 way to unlock 0 computers
        
        for (int i = 1; i <= n; i++) {
            long count = 0;
            for (int j = 0; j < i; j++) {
                if (complexity[j] < complexity[i - 1]) {
                    count = (count + dp[j]) % MOD;
                }
            }
            dp[i] = count * factorial(i - 1) % MOD;
        }
        
        return (int) dp[n];
    }

    private long factorial(int x) {
        long result = 1;
        for (int i = 2; i <= x; i++) {
            result = (result * i) % MOD;
        }
        return result;
    }
}